src/exception.cr

```diff
@@ -163,3 +163,15 @@ class NilAssertionError < Exception
     super(message)
   end
 end
+
+# Raised when trying to modify an object that is read-only.
+#
+# ```
+# slice = Slice.new(10, read_only: true) { |i| i }
+# slice[0] = 1
+# ```
+class ReadOnlyError < Exception
+  def initialize(message)
+    super(message)
+  end
```

Comment:
Why defining an `initialize` here?

Reply:
Because the message is optional in `Exception`. I want to force users of this exception to provide why the type was set to read-only.

Reply:
If so, then it needs to be defined with `: String` argument restriction, otherwise you could pass a `nil` as well.

Reply:
This error is supposed to be re-used elsewhere? It may be a better idea to have a dedicated error for reading a read-only header.

Reply:
Having both will be the best, to rescue only headers errors, or only read-only errors - or both.

Reply:
Why? I can't imagine a scenario where both can happen. And this exception should not be rescued, it means a bug in your code.

Reply:
Sure here, why using a global `ReadOnly` error then?
I don't think all `ReadOnly` exceptions (if any) would be on this case.

Just create a dedicated error where it is used; in the `HTTP::Headers` module - more importantly if not planned to be reused elsewhere.

Reply:
Because the standard library shouldn't be raising `Exception`, it should be raising specific errors. Slice was already doing that and I thought about using the same exception type because the reason is the same: an object was set to read-only mode.

Reply:
Sure, just wondering about where to place this very error, to provide more context, which will prevent to specify `HTTP::Headers` in the message (and just say `headers`)...
That's not a big deal then.

Reply:
I was thinking of something like `Cannot modify headers (HTTP::Headers::ReadOnlyError)` instead of `HTTP::Headers are in read-only (ReadOnlyError)`.

Reply:
Just my opinion, but I don't think discussing these things matter.